dynamic_programming/factorial.py

```diff
@@ -0,0 +1,35 @@
+#Factorial of a number using memoization
+def factorial(num):
+    """
+    >>> factorial(7)
+    5040
+    >>> factorial(-1)
+    'Number should not be negative.'
+    >>> [factorial(i) for i in range(5)]
+    [1, 1, 2, 6, 24]
+    """
+    
+    if num<0:
+        return "Number should not be negative."
+    if result[num]!=-1:
+        return result[num]
+    else:
+        result[num]=num*factorial(num-1)
+        #uncomment the following to see how recalculations are avoided
+        #print(result)
+        return result[num]
+
+#factorial of num
+#uncomment the following to see how recalculations are avoided
+##result=[-1]*10
+##result[0]=result[1]=1
+##print(factorial(5))
+# print(factorial(3))
+# print(factorial(7))
+
+
+if __name__ == "__main__":
+    import doctest
+    result=[-1]*10
+    result[0]=result[1]=1
```

Comment:
Hint: Move lines 33 and 34 ___inside___ the __factorial()__ function.

Reply:
Ok but if I do that the values are recomputed every time the function is called i.e. I don't want to recompute 3! if 7! is already computed.

Reply:
Understood.  Then put those two lines at the __top__ of the file at _global scope_ __before the function and the __main()__

Reply:
Thank You!